src/Mapbender/CoreBundle/Component/Base/ConfigurationBase.php

```diff
@@ -0,0 +1,105 @@
+ */
+abstract class ConfigurationBase
+{
+    protected static $classDefaults = array();
+
+    /**
+     * Convert the instance to an array representation.
+     *
+     * @return mixed[]
+     */
+    abstract public function toArray();
+
+    /**
+     * Factory method. Instantiates class and prepopulates it with values from given $options.
+     *
+     * @param mixed[] $options
+     * @param boolean $strict to throw an Exception if unrecognized options were passed
+     * @return static
+     * @throws \InvalidArgumentException if $options is not an array
+     * @throws \RuntimeException if unrecognized keys are found in $options
+     */
+    public static function fromArray($options, $strict = true)
+    {
+        if (!is_array($options)) {
+            if ($strict) {
+                throw new \InvalidArgumentException('Options must be an array, is: ' . print_r($options, true));
+            } else {
+                $options = array();
+            }
+        }
+        $instance = new static();
+        $instance->populateAttributes($options, $strict);
+        return $instance;
+    }
+
+    /**
+     * Perform mass attribute population from array via magic attribute access.
+     *
+     * @param mixed[] $options
+     * @param boolean $strict
+     * @throws \RuntimeException if unsupported value keys are found and $strict == true
+     */
+    protected function populateAttributes($options, $strict)
+    {
+        if ($strict) {
+            $validateAgainst = $this->validSet();
+            $badKeys = array_keys(array_diff_key($options, $validateAgainst));
+            if ($badKeys) {
+                $message = "Unsupported " . get_class($this) . " keys in options: " . implode(", ", $badKeys);
+                $message .= "; have: " . implode(", ", array_keys($validateAgainst));
+                throw new \RuntimeException($message);
+            }
+        }
+        $remap = $this->keyToAttributeMapping();
+        foreach ($remap as $arrayKey =>  $attributeName) {
+            if (isset($options[$arrayKey])) {
+                $options[$attributeName] = $options[$arrayKey];
+                unset($options[$arrayKey]);
+            }
+        }
+        foreach ($options as $key => $value) {
+            $this->{$key} = $value;
+        }
+    }
+
+    /**
+     * Child classes should override this if attribute names diverge from keys used in array representations.
+     * Return a mapping of arrayKey => attributeName
+     *
+     * @return string[]
+     */
+    protected static function keyToAttributeMapping()
+    {
+        return array();
+    }
+
+    /**
+     * @return array
+     * Policy; no "get" prefix for non-serializible getter functions in objects that are persisted
+     * or exported or otherwise magically scanned for methods with "get" / "is" / "has" etc
+     */
+    public static function validSet()
+    {
+        $cls = get_called_class();
```

Comment:
Maybe a better name for $cls, like $calledClass or so?

Reply:
This whole class structure turned out to be transient, i.e. it helped me understand (and unit-test) what was going on. It's a candidate for deprecation + removal now, along with the WmsInstanceConfiguration and WmsInstanceConfigurationOptions.